tony-core/src/main/java/com/linkedin/tony/TonyApplicationMaster.java

```diff
@@ -368,12 +380,34 @@ private boolean run(String[] args) {
     stop();
     long completed = System.currentTimeMillis();
     printTaskUrls();
+    Event finishEvent = constructEvent("APPLICATION_FINISHED");
+    eventQueue.add(finishEvent);
     // By this time jobDir should have been set
     HistoryFileUtils.createHistoryFile(fs,
-        TonyJobMetadata.newInstance(yarnConf, appIdString, started, completed, succeeded), jobDir);
+        TonyJobMetadata.newInstance(yarnConf, appIdString, started, completed, succeeded), jobDir, eventQueue,
+        writeRetryCnt);
     return succeeded;
   }
 
+  private Event constructEvent(String type) {
```

Comment:
I think event construction should actually be done separately for each event, since each event may require custom parameters. E.g.: TASK_INITED would need the task id but APPLICATION_INITED does not.

MRAppMaster just directly does

```
Event event = new XEvent(...);
dispatcher.getEventHandler().handler(event);
```